k8s/apis/workflow/v1/activation_webhook.go

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ * SPDX-License-Identifier: MIT
+ */
+
+package v1
+
+import (
+	"context"
+	"gopls-workspace/apis/metrics/v1"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// log is for logging in this package.
+var activationlog = logf.Log.WithName("activation-resource")
+var myActivationClient client.Client
+var catalogWebhookValidationMetrics *metrics.Metrics
+
+func (r *Activation) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	myActivationClient = mgr.GetClient()
+	mgr.GetFieldIndexer().IndexField(context.Background(), &Activation{}, ".metadata.name", func(rawObj client.Object) []string {
+		activation := rawObj.(*Activation)
+		return []string{activation.Name}
+	})
+
+	// initialize the controller operation metrics
+	if catalogWebhookValidationMetrics == nil {
+		metrics, err := metrics.New()
+		if err != nil {
+			return err
+		}
+		catalogWebhookValidationMetrics = metrics
+	}
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-workflow-symphony-v1-activation,mutating=true,failurePolicy=fail,sideEffects=None,groups=workflow.symphony,resources=activations,verbs=create;update,versions=v1,name=mactivation.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &Activation{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *Activation) Default() {
+	activationlog.Info("default", "name", r.Name)
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+
+//+kubebuilder:webhook:path=/validate-workflow-symphony-v1-activation,mutating=false,failurePolicy=fail,sideEffects=None,groups=workflow.symphony,resources=activations,verbs=create;update,versions=v1,name=mactivation.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &Activation{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Activation) ValidateCreate() (admission.Warnings, error) {
```

Comment:
Good change. Can we also extend scenario or k8s unit test to check the webhook? e.g. You can try to create an activation whose campaign doesn't exist. And check the expected error message.

Reply:
Currently the webhook is manually tested. Discuss with Jiawei offline and will add a test in a separate PR.